osquery/tables/applications/linux/docker_deb_packages.cpp

```diff
@@ -0,0 +1,196 @@
+
+static constexpr size_t BUF_SIZE{65536};
+
+void extractDebPackageInfo(const struct pkginfo* pkg,
+                           rapidjson::Writer<rapidjson::StringBuffer>& w) {
+  struct varbuf vb;
+  varbuf_init(&vb, 20);
+  w.StartObject();
+
+  // Iterate over the desired fieldinfos, calling their fwritefunctions
+  // to extract the package's information.
+  const struct fieldinfo* fip = nullptr;
+  for (fip = fieldinfos; fip->name; fip++) {
+    fip->wcall(&vb, pkg, &pkg->installed, fw_printheader, fip);
+
+    std::string line = vb.string();
+    if (!line.empty()) {
+      size_t separator_position = line.find(':');
+
+      std::string key = line.substr(0, separator_position);
+      std::string value = line.substr(separator_position + 1, line.length());
+
+      auto it = kFieldMappings.find(key);
+      if (it != kFieldMappings.end()) {
+        boost::algorithm::trim(value);
+        w.Key(it->second.c_str());
+        w.String(std::move(value));
+      }
+    }
+    varbuf_reset(&vb);
+  }
+  varbuf_destroy(&vb);
+
+  w.EndObject();
+}
+
+void getDebPackages(int fd) {
+  if (!osquery::isDirectory(kDPKGPath)) {
+    TLOG << "Cannot find DPKG database: " << kDPKGPath;
+    return;
+  }
+
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  writer.StartArray();
+
+  struct pkg_array packages;
+  dpkg_setup(&packages);
+
+  for (int i = 0; i < packages.n_pkgs; i++) {
+    struct pkginfo* pkg = packages.pkgs[i];
+    // Casted to int to allow the older enums that were embeded in the
+    // packages
+    // struct to be compared
+    if (static_cast<int>(pkg->status) ==
+        static_cast<int>(PKG_STAT_NOTINSTALLED)) {
+      continue;
+    }
+
+    extractDebPackageInfo(pkg, writer);
+  }
+
+  dpkg_teardown(&packages);
+  writer.EndArray();
+
+  size_t bufSize = buffer.GetSize();
+  if (bufSize < 1) {
+    return;
+  }
+  if (write(fd, buffer.GetString(), bufSize) != (ssize_t)bufSize) {
+    VLOG(1) << "error writing pkginfo ";
+  }
+}
+
+QueryData genDockerDebPackages(QueryContext& context) {
+  QueryData results;
+
+  for (const auto& id : context.constraints["id"].getAll(EQUALS)) {
+    if (!checkConstraintValue(id)) {
+      continue;
+    }
+
+    pt::ptree container;
+    Status s = dockerApi("/containers/" + id + "/json", container);
+    if (!s.ok()) {
+      VLOG(1) << "Error getting docker container info " << id << ": "
+              << s.what();
+      continue;
+    }
+    if (!container.get<bool>("State.Running", false)) {
+      continue;
+    }
+    pid_t pid = container.get<int>("State.Pid", 0);
+    if (pid == 0) {
+      VLOG(1) << "Error getting pid for container " << id;
+      continue;
+    }
+
+    int fds[2];
+    if (socketpair(AF_UNIX, SOCK_STREAM, 0, fds) == -1) {
+      VLOG(1) << "unable to open socket pair" << strerror(errno);
+      continue;
+    }
+
+    NamespaceOps nsOps(pid, fds[1]);
+    Status forkOps = nsOps.invoke(getDebPackages);
```

Comment:
@diptanu How is this better (or possibly worse) than container cp (get file) option? /var/lib/dpkg/status seem to contain all the necessary information!

Reply:
Even if the parsing of the `/var/lib/dpkg/status` file is done outside the container, it's still worth doing it in a process with reduced privileges. Partly because the dpkg library parser is not hardened, so manipulating the file could provide a path to privilege escalation/container escape.

Reply:
@spasam This implementation could be extended in the future to any other container implementation, and not just Docker. Also, even if we just wanted to do docker, copying files to a temp location every time we have to do this sounds messy. Also, not sure how that would look like for other package managers like RPM which behind the scenes use the Berkley DB, we will have to copy the entire database outside the container.

Reply:
@diptanu Thanks. If the plan is to supporting additional containers etc, wouldn't it be better to rename: docker_deb_packages -> container_packages (one table for both deb/rpm etc and not be tied to docker)

I haven't spent time looking into the details of nsOp/fork. What happens to connections to TLS server, currently executing queries, handles to glog/rocksdb etc? This would pose problems, if the forked process starts reporting query results etc.

Reply:
We are not doing any form of exec or subprocess here, so I don't see how the process can have any un-controlled behaviour such as the one you are describing? But, it would be good if you can spend some time looking at the code and think about possible thread models.

On the topic of extending the logic to other containers, I don't intend to implement this for anything other than Docker right now, but I will add support for reporting rpm packages soon. I meant to say that, this construct of using a `NamespaceOps` to introspect other namespaces could be used if we decided to add support for other types of containers in the future. The logic for each container types would be slightly different, like in the case of Docker we use the docker api to first get the main pid of the container, for LXC we will have to use liblxc or something similar to get the parent pid, etc.